compiler/src/regular-grammar/index.ts

```diff
@@ -283,75 +354,291 @@ class RegularGrammar extends CompilerClass {
+    const stateQueue: Set<string>[] = [];
+    const processedStates: Set<string>[] = [];
+    const DEAD_STATE = 'Φ';
+
+    // Get the alphabet(List of terminals)
+    for (const state in nfa) {
+      for (const symbol in nfa[state].nodes) {
+        alphabet.add(symbol);
+      }
+    }
+
+    dfa[DEAD_STATE] = { nodes: {}, final: false };
+    for (const symbol of alphabet) {
+      dfa[DEAD_STATE].nodes[symbol] = new Set([DEAD_STATE]);
+    }
+
+    // Helper function to get the next state set
+    const getNextStateSet = (
+      currentSet: Set<string>,
+      symbol: string
+    ): Set<string> => {
+      const nextSet = new Set<string>();
+      for (const state of currentSet) {
+        if (nfa[state].nodes[symbol]) {
+          nfa[state].nodes[symbol].forEach((nextState) =>
+            nextSet.add(nextState)
+          );
+        }
+      }
+      return nextSet;
+    };
+
+    // Helper function to convert set to sorted string (for state naming)
+    const setToString = (set: Set<string>): string =>
+      Array.from(set).sort().join(',');
+
+    // Start with the initial state
+    const initialState = new Set(['S']);
+    stateQueue.push(initialState);
+
+    while (stateQueue.length > 0) {
+      const currentSet = stateQueue.shift()!;
+      const currentState = setToString(currentSet);
+
+      if (processedStates.some((set) => setToString(set) === currentState))
+        continue;
+
+      processedStates.push(currentSet);
+
+      dfa[currentState] = { nodes: {}, final: false };
+
+      // Check if the current state set contains any final state from NFA
+      dfa[currentState].final = Array.from(currentSet).some(
+        (state) => nfa[state].final
+      );
+
+      for (const symbol of alphabet) {
+        const nextSet = getNextStateSet(currentSet, symbol);
+        if (nextSet.size > 0) {
+          const nextState = setToString(nextSet);
+          dfa[currentState].nodes[symbol] = new Set([nextState]);
+
+          if (!processedStates.some((set) => setToString(set) === nextState)) {
+            stateQueue.push(nextSet);
+          }
+        } else {
+          // Transition to dead state if next set is empty
+          dfa[currentState].nodes[symbol] = new Set([DEAD_STATE]);
+        }
+      }
+    }
+
+    this.result = dfa;
+    return this;
+  }
+  minimizeDFA() {
+    const dfa: FAGraph = this.toDFA().result;
+    const alphabet: Set<string> = new Set();
+    const result: FAGraph = {};
+
+    // Get the alphabet
+    for (const state in dfa) {
+      for (const symbol in dfa[state].nodes) {
+        alphabet.add(symbol);
+      }
+    }
+
+    const finalStates = new Set(
+      Object.keys(dfa).filter((state) => dfa[state].final)
+    );
+    const nonFinalStates = new Set(
+      Object.keys(dfa).filter((state) => !dfa[state].final)
+    );
+
+    let partitions = [finalStates, nonFinalStates].filter(
+      (set) => set.size > 0
+    );
+    let workList = [...partitions];
```

Comment:
```suggestion
    const partitions = [finalStates, nonFinalStates].filter(
      (set) => set.size > 0
    );
    const workList = [...partitions];
```
Small nit, both are never re-assigned, so let's keep them as `const`